### PR TITLE
# 🚀 feat(chat): improve error handling and add retry logic for chat stream

### PR DESCRIPTION
**Pull request by Andi-IM**
## 📋 **Deskripsi**

Pull request ini mengimplementasikan perbaikan komprehensif pada sistem AI Chat untuk meningkatkan stabilitas koneksi, penanganan error, dan pengalaman pengguna secara keseluruhan.

**Closes:** #52 - Peningkatan Stabilitas dan UX pada Fitur AI Chat

### 🔄 **Retry Logic dengan Exponential Backoff**
- ✅ Menambahkan mekanisme retry otomatis untuk koneksi chat stream yang gagal
- ✅ Implementasi exponential backoff (2^n detik) untuk mencegah spam request  
- ✅ Maksimal 3 kali percobaan ulang sebelum menampilkan error final
- ✅ State management untuk tracking retry attempts

### 🛡️ **Enhanced Error Handling**
- ✅ Penanganan khusus untuk `RealtimeSubscribeException` pada repository layer
- ✅ Error tracking terintegrasi dengan AnalyticsManager
- ✅ Pesan error yang user-friendly dalam bahasa Indonesia
- ✅ Graceful degradation saat koneksi bermasalah

### 📊 **Performance Monitoring**
- ✅ Tracking metrics untuk screen rendering performance
- ✅ Monitoring durasi build dan raster untuk optimasi UI
- ✅ Trace analytics untuk operasi chat stream
- ✅ Method `calculateScreenRendering()` untuk performance insights

### 🎯 **UX Improvements**
- ✅ Wrap chat screen dengan `GestureDetector` untuk dismiss keyboard
- ✅ Loading states yang lebih informatif
- ✅ Visual feedback untuk retry attempts
- ✅ Error messages yang actionable
        
#### 📁 `lib/data/repositories/brick/chat_repository_supabase_impl.dart`
```diff
+ } on RealtimeSubscribeException catch (e, st) {
+   dev.log('Realtime subscription error: ${e.toString()}', 
+           name: 'Chat Repository', stackTrace: st);
+   throw ConnectionFailure('Realtime connection failed: ${e.toString()}');
```
- Menambahkan penanganan khusus untuk RealtimeSubscribeException
- Logging yang lebih detail untuk debugging

#### 📁 `lib/presentation/ai_chat/provider/ai_chat_notifier.dart`
**State Management Baru:**
```dart
+ bool _isRetrying = false;
+ int _retryCount = 0;
+ static const int _maxRetries = 3;
+ String? _errorMessage;
```

**Method Baru:**
- `retryConnection()` - Manual retry trigger
- `calculateScreenRendering()` - Performance tracking
- Enhanced `initStream()` dengan error handling dan retry logic

#### 📁 `lib/presentation/ai_chat/screens/ai_chat_screen.dart`
```diff
+ body: GestureDetector(
+   onTap: () => FocusScope.of(context).unfocus(),
+   child: Consumer<AiChatNotifier>(
```
- Wrap dengan GestureDetector untuk UX improvement

### **Algoritma Retry Logic:**
1. **Initial Connection**: Attempt koneksi pertama
2. **Error Detection**: Catch dan log error dengan AnalyticsManager
3. **Retry Decision**: Check apakah `_retryCount < _maxRetries`
4. **Exponential Backoff**: Delay = 2^retryCount seconds
5. **User Feedback**: Update UI dengan status retry
6. **Max Retries**: Tampilkan error final jika gagal semua

### **Manual Testing Scenarios:**
- [ ] **Koneksi Normal**: Chat berfungsi normal tanpa error
- [ ] **Koneksi Terputus**: Retry mechanism aktif dengan feedback yang tepat
- [ ] **Koneksi Lambat**: Exponential backoff bekerja dengan baik
- [ ] **Keyboard Dismiss**: Tap di luar input field menutup keyboard
- [ ] **Performance**: Tidak ada lag saat rendering chat messages

### **Error Scenarios Tested:**
- [ ] `RealtimeSubscribeException` handling
- [ ] Network timeout scenarios  
- [ ] Multiple rapid disconnections
- [ ] Max retry limit reached
        
### **Development Checklist:**
- [x] Code mengikuti style guide project
- [x] Semua method memiliki dokumentasi yang memadai
- [x] Error handling comprehensive dengan logging
- [x] State management yang clean dan predictable
- [x] Performance monitoring terintegrasi
- [x] User feedback yang informatif

### **Testing Checklist:**
- [ ] Unit tests untuk retry logic
- [ ] Integration tests untuk chat stream
- [ ] Manual testing pada berbagai kondisi network
- [ ] Performance testing untuk screen rendering
- [ ] Error scenario testing

### **Code Quality:**
- [x] No breaking changes
- [x] Backward compatibility maintained
- [x] Memory leaks prevention (proper stream disposal)
- [x] Analytics tracking implemented
- [x] Internationalization ready (Indonesian messages)

### **Before vs After:**

| Aspek | Before | After |
|-------|--------|-------|
| **Error Handling** | Basic try-catch | Comprehensive dengan retry |
| **User Feedback** | Minimal | Informative dengan progress |
| **Connection Stability** | Single attempt | Auto-retry dengan backoff |
| **Performance Monitoring** | None | Integrated analytics |
| **UX** | Basic | Enhanced dengan keyboard dismiss |

### **Metrics to Monitor:**
- Chat connection success rate
- Average retry attempts per session
- Screen rendering performance
- User engagement with retry feature

### **Focus Areas untuk Reviewer:**
1. **Error Handling Logic**: Pastikan semua edge cases tertangani
2. **State Management**: Verify state transitions yang benar
3. **Performance Impact**: Check tidak ada performance regression
4. **User Experience**: Test flow retry dan error messages
5. **Code Quality**: Review clean code principles

### **Testing Recommendations:**
```bash
# Test dengan koneksi tidak stabil
flutter run --debug
# Matikan/nyalakan WiFi berulang kali
# Observe retry behavior dan user feedback
```

## 📝 **Breaking Changes**
**None** - Semua perubahan backward compatible

## 🚀 **Deployment Notes**
- Tidak ada migration database diperlukan
- Tidak ada environment variable baru
- Analytics events baru akan mulai tracked setelah deploy

### CODE REVIEW

File: lib/presentation/tourism_spot/pages/tourism_spot_page.dart [hunk 1]
```diff
@@ -189,8 +189,11 @@ class _TourismSpotPageState extends State<TourismSpotPage> {
           Expanded(
             child: Consumer<TourismSpotNotifier>(
               builder: (context, notifier, _) {
+                final TextEditingController controller = TextEditingController(
+                  text: notifier.searchQuery,
+                );
                 return TextField(
-                  controller: TextEditingController(text: notifier.searchQuery),
+                  controller: controller,
```
[the remark this answers is not in the document]
Comment (Andi-IM) [reply lib/presentation/tourism_spot/pages/tourism_spot_page.dart]: di luar scope pr ini

File: lib/data/repositories/brick/chat_repository_supabase_impl.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+class ChatRepositoryImpl implements ChatRepository {
+  final SupabaseClient _client;
+
+  ChatRepositoryImpl() : _client = Supabase.instance.client;
+
+  @override
+  Future<void> clearHistory() async {
+    try {
+      final user = _client.auth.currentUser;
+      if (user == null) throw ServerFailure('User not logged in.');
```
[the remark this answers is not in the document]
Comment (Andi-IM) [reply lib/data/repositories/brick/chat_repository_supabase_impl.dart]: OK

File: lib/presentation/ai_chat/provider/ai_chat_notifier.dart [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+               initStream(); // Retry connection
+             }
+           });
+         } else {
+           // Max retries reached
+           _isRetrying = false;
+           _errorMessage = 'Koneksi gagal. Silakan periksa koneksi internet Anda dan coba lagi.';
+           notifyListeners();
+         }
+       },
```
[the remark this answers is not in the document]
Comment (Andi-IM) [reply lib/presentation/ai_chat/provider/ai_chat_notifier.dart]: OK

File: lib/presentation/ai_chat/provider/ai_chat_notifier.dart [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        _errorMessage = 'Failed to generate response.';
+        return;
+      }
+      _manager.setTraceMetric('ResponseLength', 'Chars', responseText.length);
+      _manager.stopTrace('Generate Response');
+
+      // store response
+      _manager.startTrace('Store Response');
+      await _repository.storeChat(userMessageText, true);
+      await _repository.storeChat(responseText, false);
```
[the remark this answers is not in the document]
Comment (Andi-IM) [reply lib/presentation/ai_chat/provider/ai_chat_notifier.dart]: OK
